README.txt
= osxscreenshot

* http://github.com/topfunky/osxscreenshot

== DESCRIPTION:

A Ruby wrapper around webkit2png.py to easily and programmatically
capture screenshots of websites, then crop and resize them. Mac OS X
only.

== FEATURES/PROBLEMS:

* Uses the Python built-in to Mac OS X.
* You may be able to use snapurl instead of webkit2png.py: http://gemcutter.org/gems/snapurl
* See examples in use at http://blog.peepcode.com/archives and http://news.peepcode.com/

== REQUIREMENTS:

* Uses Python's Cocoa support (standard in Mac OS X). 
* Uses ImageMagick's command-line tools to resize the image (mogrify). See http://github.com/masterkain/ImageMagick-sl. I hope to replace this with MacRuby or RubyCocoa instead.

== INSTALL:

* gem install osxscreenshot

== DEVELOPERS:

Call the +capture+ method which returns the path to a tempfile containing the image:

    file_path = OSX::Screenshot.capture("http://peepcode.com")
    # => "/tmp/20293-202020293-2020-full.png" at 320x480

Or, pass some (optional) options.

    file_path =
      OSX::Screenshot.capture(my_url, {
                         :tmpdir     => "#{Sinatra::Application.root}/tmp",
                         :webkit2png => "#{Sinatra::Application.root}/bin/webkit2png.py",
                         :mogrify    => "/opt/local/bin/mogrify",
                         :width      => 220,
                         :height     => 270
                       })
    system "mv #{file_path} #{permanent_path}"

== LICENSE:

(The MIT License)

Copyright (c) 2010 Topfunky Corporation

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 0.0.5 / 2010-05-10

* Wait for killed processes so zombies don't form

=== 0.0.4 / 2010-02-19

* Kill pid properly on timeout
* Throw OSX::Screenshot::CommandFailed if any command returns nonzero.

=== 0.0.2 / 2010-02-16

* Added simple usage example.
* Minor URL changes.
* Noted ImageMagick/mogrify command-line requirement.

=== 0.0.1 / 2010-02-16

* First release




test/test_osxscreenshot.rb
require "test/unit"
require "osxscreenshot"

class TestOsxscreenshot < Test::Unit::TestCase

  def test_loads_url
    @tmpfile = OSX::Screenshot.capture("http://example.com")
    assert_not_nil @tmpfile
    assert File.exist?(@tmpfile)
  end

  def test_loads_tall_url
    @tmpfile = OSX::Screenshot.capture("http://blog.peepcode.com/archives")

    extra_file_name = @tmpfile.gsub(/-0\.png$/, "-1.png")
    assert !File.exist?(extra_file_name)
  end

  def test_uses_custom_tmpdir
    @tmpfile = OSX::Screenshot.capture("http://example.com", :tmpdir => "./tmp")
    assert_match(/^\.\/tmp/, @tmpfile)
    assert File.exist?(@tmpfile)
  end

  def test_handles_timeout
    @tmpfile = OSX::Screenshot.capture("http://example.com", {
                                         :timeout => 1,
                                         :webkit2png => "sleep 5 &&"
                                       })
    assert_nil @tmpfile
  end

  def test_handles_error_when_exiting
    @tmpfile = nil
    assert_raise OSX::Screenshot::CommandFailed do
      OSX::Screenshot.capture("http://example.com", {
                                :webkit2png => "exit 1;"
                              })
    end
  end

  def teardown
    if @tmpfile
      FileUtils.rm @tmpfile
    end
  end

end


lib/osxscreenshot.rb
require 'open4'

module OSX
  class Screenshot
    VERSION = '0.0.5'
    
    class CommandFailed < StandardError; end
    
    ##
    # Takes a screenshot of a website, optionally resizes it, and writes
    # it to a temporary directory. Returns the file path to the
    # temporary file.
    #
    # It's assumed that you will move the file to a permanent
    # destination or store it in a database.
    #
    # Options include:
    #
    # tmpdir::     Path to tmp directory where files will be stored
    # webkit2png:: Path to webkit2png.py command
    # mogrify::    Path to mogrify command
    #
    # Usage:
    #
    #   output_screenshot_path =
    #     OSX::Screenshot.capture(article_url, {
    #                        :tmpdir     => "#{Sinatra::Application.root}/tmp",
    #                        :webkit2png => "#{Sinatra::Application.root}/bin/webkit2png.py",
    #                        :mogrify    => "/opt/local/bin/mogrify",
    #                        :width      => 220,
    #                        :height     => 270
    #                      })
    #   system "mv #{output_screenshot_path} #{local_path}"

    def self.capture(url, options={})
      obj = new(url, options)
      obj.capture
    end

    def initialize(url, options)
      vendored_webkit2png = File.expand_path(File.join(File.dirname(__FILE__),
                                                       "..",
                                                       "vendor",
                                                       "webkit2png.py"))
      @url = url
      @options = {
        :tmpdir     => "/tmp",
        :webkit2png => vendored_webkit2png,
        :mogrify    => "mogrify",
        :width      => 320,
        :height     => 480,
        :timeout    => 30
      }.merge(options)
    end

    def capture
      random_id        = [@url.length, Time.now.to_i.to_s, rand(10000)].join('-')
      tmp_abs_filename = File.join(@options[:tmpdir], "#{random_id}-full.png")
      tmp_dir          = File.dirname(tmp_abs_filename)
      FileUtils.mkdir_p(tmp_dir)

      webkit2png_command = [@options[:webkit2png],
                            "--full",
                            "--filename",    random_id,
                            "--dir",         @options[:tmpdir],
                            @url
                           ].join(' ')

      run_command(webkit2png_command) do
        return resize(tmp_abs_filename, @options[:width], @options[:height])
      end
    end

    def resize(tmp_abs_filename, width, height)
      # Example: mogrify -resize 320x peepcodecom-full.png -crop 320x480 peepcodecom-full.png

      mogrify_command = [@options[:mogrify],
                         "-resize", "#{width}x",
                         tmp_abs_filename,
                         "-crop",   "#{width}x#{height}",
                         tmp_abs_filename
                        ].join(' ')
      run_command(mogrify_command) do
        output_filename = if File.exist?(tmp_abs_filename)
                            # Add full width and height to image
                            extent_command = [@options[:mogrify],
                                              "-extent", "#{width}x#{height}",
                                              tmp_abs_filename].join(' ')
                            run_command(extent_command) { return tmp_abs_filename }
                          elsif File.exist?(tmp_abs_filename.gsub(/\.png/, '-0.png'))
                            # Remove extra file generated by cropping.
                            FileUtils.rm(tmp_abs_filename.gsub(/\.png/, '-1.png'))
                            tmp_abs_filename.gsub(/\.png/, '-0.png')
                          end
        return output_filename
      end
    end

    def run_command(cmd, &block)
      pid = nil
      status = nil

      Timeout.timeout(@options[:timeout]) {
        pid, i, o, e = open4(cmd)
        ignored, status = Process::waitpid2 pid
      }
      
      # Success
      if (status.exitstatus == 0)
        return yield
      end
      
      # Error
      raise CommandFailed, "Command failed: #{cmd}"
    rescue Timeout::Error => e
      Process.kill "KILL", pid
      Process.wait pid
      return nil
    end

  end
end


